native/core/src/execution/shuffle/codec.rs

```diff
@@ -0,0 +1,694 @@
+                // write null buffer
+                self.write_null_buffer(arr.nulls())?;
+            }
+            DataType::Int16 => {
+                let arr = col.as_any().downcast_ref::<Int16Array>().unwrap();
+                // write data buffer
+                self.write_buffer(arr.values().inner())?;
+                // write null buffer
+                self.write_null_buffer(arr.nulls())?;
+            }
+            DataType::Int32 => {
+                let arr = col.as_any().downcast_ref::<Int32Array>().unwrap();
+                // write data buffer
+                self.write_buffer(arr.values().inner())?;
+                // write null buffer
+                self.write_null_buffer(arr.nulls())?;
+            }
+            DataType::Int64 => {
+                let arr = col.as_any().downcast_ref::<Int64Array>().unwrap();
+                // write data buffer
+                self.write_buffer(arr.values().inner())?;
+                // write null buffer
+                self.write_null_buffer(arr.nulls())?;
+            }
+            DataType::Float32 => {
+                let arr = col.as_any().downcast_ref::<Float32Array>().unwrap();
+                // write data buffer
+                self.write_buffer(arr.values().inner())?;
+                // write null buffer
+                self.write_null_buffer(arr.nulls())?;
+            }
+            DataType::Float64 => {
+                let arr = col.as_any().downcast_ref::<Float64Array>().unwrap();
+                // write data buffer
+                self.write_buffer(arr.values().inner())?;
+                // write null buffer
+                self.write_null_buffer(arr.nulls())?;
+            }
+            DataType::Date32 => {
+                let arr = col.as_any().downcast_ref::<Date32Array>().unwrap();
+                // write data buffer
+                self.write_buffer(arr.values().inner())?;
+                // write null buffer
+                self.write_null_buffer(arr.nulls())?;
+            }
+            DataType::Timestamp(TimeUnit::Microsecond, _) => {
+                let arr = col
+                    .as_any()
+                    .downcast_ref::<TimestampMicrosecondArray>()
+                    .unwrap();
+                // write data buffer
+                self.write_buffer(arr.values().inner())?;
+                // write null buffer
+                self.write_null_buffer(arr.nulls())?;
+            }
+            DataType::Decimal128(_, _) => {
+                let arr = col.as_any().downcast_ref::<Decimal128Array>().unwrap();
+                // write data buffer
+                self.write_buffer(arr.values().inner())?;
+                // write null buffer
+                self.write_null_buffer(arr.nulls())?;
+            }
+            DataType::Utf8 => {
+                let arr = col.as_any().downcast_ref::<StringArray>().unwrap();
+                // write data buffer
+                self.write_buffer(arr.values())?;
```

Comment:
FYI if the array is sliced this will write the full data buffer, as the slice is not "materialized" into the buffer